pkgs/applications/misc/gopass-ui/default.nix

```diff
@@ -0,0 +1,132 @@
+{ lib
+, system
+, stdenv
+, fetchgit
+, fetchurl
+, fetchFromGitHub
+, makeDesktopItem
+, makeWrapper
+, runCommand
+, writeTextFile
+, electron_7
+, electron-chromedriver_3 ? null
+, gifsicle
+, jq
+, libwebp
+, mozjpeg
+, nix-gitignore
+, nodejs-12_x
+, nodePackages
+, optipng
+, p7zip
+, pngquant
+, python2
+, utillinux
+, darwin
+, xcbuild
+, bash
+} @ args:
+
+stdenv.mkDerivation rec {
+  pname = "gopass-ui";
+  version = "0.7.0";
+  name = "${pname}-${version}";
+
+  nativeBuildInputs = [
+    makeWrapper
+    nodejs-12_x
+  ];
+
+  src = fetchFromGitHub {
+    owner  = "codecentric";
+    repo   = "gopass-ui";
+    rev    = "v${version}";
+    sha256 = "1yn8x3raj8hqc6jshqdbn6yfkh6qmp2j5i6b3xz8q02xyvb1r241";
+  };
+
+  nodeDeps = (import ./node-dependencies.nix {
+      inherit system;
+      pkgs = args // { inherit electron-chromedriver_3; };
+      gopass-ui-src = src;
+    }).package;
+
+  buildInputs = [
+    electron_7
+  ] ++ stdenv.lib.optionals stdenv.isDarwin [ xcbuild darwin.DarwinTools ];
+
+  buildPhase = ''
+    dev_node_modules="${nodeDeps}/lib/node_modules/gopass-ui/node_modules"
+    ln -s -t "." "$dev_node_modules"
+
+    export PATH="$dev_node_modules/.bin:$PATH"
+
+    # npm run build
+    NODE_ENV=production webpack --config webpack.main.prod.config.js
+    NODE_ENV=production webpack --config webpack.renderer.explorer.prod.config.js
+    NODE_ENV=production webpack --config webpack.renderer.search.prod.config.js
+  '';
+
+  doCheck = true;
+
+  checkPhase = ''
+    # npm run release:check
+    tslint '{src,test,mocks}/**/*.{ts,tsx,js,jsx}' --project ./tsconfig.json
+    jest --testRegex '\.test\.tsx?$'
+
+    # npm run test:integration
+    # jest --testRegex '\.itest\.ts$'
+  '';
+
+  installPhase = ''
```

Comment:
installPhase on darwin should be different, we expect `$out/Applications/Gopass UI.app` there.

Reply:
That might be a bit more involved to fix as we're currently skipping the `electron-builder` pass which is unpure in that it attempts to download some binaries to create installable bundles. Do you know of a clean / simple way to generate the this app bundle dir content from what we have (exe and icons)?

Unpacking the official binaries bundle <https://github.com/codecentric/gopass-ui/releases/download/v0.7.0/Gopass.UI-0.7.0.dmg>, I can see there are a great deal of files there:

```bash
$ 7z x Gopass.UI-0.7.0.dmg
# ..
$ tree Gopass\ UI\ 0.7.0/Gopass\ UI.app/
Gopass UI 0.7.0/Gopass UI.app/
└── Contents
    ├── _CodeSignature
    │   └── CodeResources
    ├── Frameworks
    │   ├── Electron Framework.framework
    │   │   ├── Electron Framework
    │   │   ├── Libraries
    │   │   ├── Resources
    │   │   └── Versions
    │   │       ├── A
    │   │       │   ├── _CodeSignature
    │   │       │   │   └── CodeResources
    │   │       │   ├── Electron Framework
    │   │       │   ├── Libraries
    │   │       │   │   ├── libEGL.dylib
    │   │       │   │   ├── libffmpeg.dylib
    │   │       │   │   ├── libGLESv2.dylib
    │   │       │   │   ├── libswiftshader_libEGL.dylib
    │   │       │   │   └── libswiftshader_libGLESv2.dylib
    │   │       │   └── Resources
    │   │       │       ├── am.lproj
    │   │       │       │   └── locale.pak
# ..
    │   │       │       └── zh_TW.lproj
    │   │       │           └── locale.pak
    │   │       └── Current
    │   ├── Gopass UI Helper.app
    │   │   └── Contents
    │   │       ├── _CodeSignature
    │   │       │   └── CodeResources
    │   │       ├── Info.plist
    │   │       ├── MacOS
    │   │       │   └── Gopass UI Helper
    │   │       └── PkgInfo
    │   ├── Gopass UI Helper (GPU).app
    │   │   └── Contents
    │   │       ├── _CodeSignature
    │   │       │   └── CodeResources
    │   │       ├── Info.plist
    │   │       ├── MacOS
    │   │       │   └── Gopass UI Helper (GPU)
    │   │       └── PkgInfo
    │   ├── Gopass UI Helper (Plugin).app
    │   │   └── Contents
    │   │       ├── _CodeSignature
    │   │       │   └── CodeResources
    │   │       ├── Info.plist
    │   │       ├── MacOS
    │   │       │   └── Gopass UI Helper (Plugin)
    │   │       └── PkgInfo
    │   ├── Gopass UI Helper (Renderer).app
    │   │   └── Contents
    │   │       ├── _CodeSignature
    │   │       │   └── CodeResources
    │   │       ├── Info.plist
    │   │       ├── MacOS
    │   │       │   └── Gopass UI Helper (Renderer)
    │   │       └── PkgInfo
    │   ├── Mantle.framework
    │   │   ├── Headers
    │   │   ├── Mantle
    │   │   ├── Modules
    │   │   ├── Resources
    │   │   └── Versions
    │   │       ├── A
    │   │       │   ├── _CodeSignature
    │   │       │   │   └── CodeResources
    │   │       │   ├── Headers
    │   │       │   │   ├── Mantle.h
# ..
    │   │       │   │   └── NSValueTransformer+MTLPredefinedTransformerAdditions.h
    │   │       │   ├── Mantle
    │   │       │   ├── Modules
    │   │       │   │   └── module.modulemap
    │   │       │   └── Resources
    │   │       │       └── Info.plist
    │   │       └── Current
    │   ├── ReactiveCocoa.framework
    │   │   ├── Headers
    │   │   ├── Modules
    │   │   ├── ReactiveCocoa
    │   │   ├── Resources
    │   │   └── Versions
    │   │       ├── A
    │   │       │   ├── _CodeSignature
    │   │       │   │   └── CodeResources
    │   │       │   ├── Headers
    │   │       │   │   ├── EXTKeyPathCoding.h
# ..
    │   │       │   │   └── ReactiveCocoa.h
    │   │       │   ├── Modules
    │   │       │   │   └── module.modulemap
    │   │       │   ├── ReactiveCocoa
    │   │       │   └── Resources
    │   │       │       └── Info.plist
    │   │       └── Current
    │   └── Squirrel.framework
    │       ├── Headers
    │       ├── Modules
    │       ├── Resources
    │       ├── Squirrel
    │       └── Versions
    │           ├── A
    │           │   ├── _CodeSignature
    │           │   │   └── CodeResources
    │           │   ├── Headers
    │           │   │   ├── NSBundle+SQRLVersionExtensions.h
    │           │   │   ├── NSProcessInfo+SQRLVersionExtensions.h
    │           │   │   ├── SQRLDownloadedUpdate.h
    │           │   │   ├── SQRLUpdate.h
    │           │   │   ├── SQRLUpdater.h
    │           │   │   └── Squirrel.h
    │           │   ├── Modules
    │           │   │   └── module.modulemap
    │           │   ├── Resources
    │           │   │   ├── Info.plist
    │           │   │   └── ShipIt
    │           │   └── Squirrel
    │           └── Current
    ├── Info.plist
    ├── MacOS
    │   └── Gopass UI
    ├── PkgInfo
    └── Resources
        ├── am.lproj
        # ..
        └── zh_TW.lproj

154 directories, 194 files
```

Reply:
Also, does the executable under bin runs?

Reply:
For the electron-builder pas for mac:

```bash
$ rm -rf /home/rgauthier/.cache/electron/

$ USE_SYSTEM_MKSQUASHFS=1 electron-builder --mac --dir
  • electron-builder  version=22.5.1 os=4.19.84
  • loaded configuration  file=package.json ("build" field)
  • writing effective config  file=release/builder-effective-config.yaml
  • packaging       platform=darwin arch=x64 electron=7.2.3 appOutDir=release/mac
  • downloading     url=https://github.com/electron/electron/releases/download/v7.2.3/electron-v7.2.3-darwin-x64.zip size=64 MB parts=8
  • downloaded      url=https://github.com/electron/electron/releases/download/v7.2.3/electron-v7.2.3-darwin-x64.zip duration=15.731s
  • skipped macOS application code signing  reason=supported only on macOS

$ tree ./release/mac/Gopass\ UI.app/ | head -n 5
./release/mac/Gopass UI.app/
└── Contents
    ├── Frameworks
    │   ├── Electron Framework.framework
    │   │   ├── Electron Framework -> Versions/Current/Electron Framework
# ..
```

You can see the side effect `/home/rgauthier/.cache/electron/` of the download. electron-builder would need to be patched to avoid this. Also, this electron builder things brings some non nixpkgs electron binaries which have to be patched with patchelf. What we did instead is to use the nixpkgs `electron_7`, which prevent introducing binaries.

Reply:
We might definitely fool `electron-builder` by downloading / unpacking the things it needs under `$tmpdir/.cache/electron/` and set this temps dir as `HOME`.

Reply:
Another alternative, would be to mimick what electron-builder does (if you look at `https://github.com/electron/electron/releases/download/v7.2.3/electron-v7.2.3-darwin-x64.zip`, this is merely the content of the electron app), by initializing the app bundle dir with a tree of symlink pointing to the nixpkgs electron's own app bundle dir, then simply replace the exe part by the wrapper we already output under `$out/bin`.

Reply:
> Do you know of a clean / simple way to generate the this app bundle dir content from what we have (exe and icons)?

Unfortunatly, I don't know a good way. I've already tried to build electron based app before (keeweb - https://github.com/NixOS/nixpkgs/pull/88921), and abandoned this idea in favor of using official binaries (.dmg on darwin and .AppImage on linux).

> Also, does the executable under bin runs?

Yes, it works. So, it's already usable on darwin now.

Reply:
Indeed quite tough to get right, I have been working on this many hours, and still note quite there. I thing we might be able to do something with this `electron-builder` with some careful hacks. It might be some efforts though.

Would the current state of thing be sufficient for a initial iteration (this PR)? I think we can put the extra effort to get it perfectly right in subsequent iterations.

By curiosity, did you noticed what kind of things were missing using the current exe under bin as opposed to the full dmg release? I would assume some icon things a bit like the free desktop spec?

I shall be looking at what can be done with the `electron-builder` linux output. I think the same kind of principles should be applicable to its darwin out.

Reply:
> Would the current state of thing be sufficient for a initial iteration (this PR)?

Sure.

> By curiosity, did you notice what kind of things were missing using the current exe under bin as opposed to the full dmg release?

I noticed 2 issues now:
1. "Copy to clipboard" doesn't work (error message: "Oops, something went wrong. Please try again."), but it doesn't work with .dmg as well.
2. "Search..." form doesn't work (just can't type in), but it works with .dmg.

And I wonder why gopass-ui so slow.

Reply:
Mmh that's odd. Both 1 and 2 work fine on linux.

> And I wonder why gopass-ui so slow.

On linux, I noticed no slowness outside that this is an electron app.
It might be because the files are not bundled as an asar.

I did took some time to make the above discussed `electron-builder` pass for linux, and I thing it was a success.
It effectively create the asar bundle and copies the (already patched) electron files from the nixpkgs electron 
derivation to the output. 
I was successful in optimizing the output size by feeding it some tree of symlinks. 

If you want to take a look:  [`default.nix::mk_electron_bundle_linux`](https://github.com/NixOS/nixpkgs/pull/86169/files#diff-3a2de39ae2114b610155ba696efb9dcaR101).

Now we might attempt to create a `mk_electron_bundle_linux` in the same fashion.

Reply:
I would have a potential darwin candidate in this latest version.

See [`mk_electron_bundle_darwin`](https://github.com/NixOS/nixpkgs/pull/86169/files#diff-3a2de39ae2114b610155ba696efb9dcaR105)

I hopes it does work a real darwin machine. Tested the code as much as I could in a sandbox using an electron darwin url as input. Seems to be ok here.

Reply:
@sikmir: Do this latest incarnation with full fledged darwin packaging works any better?

Reply:
electron-builder update check failed: https://pastebin.com/i0yt1dsh

Reply:
Thanks.

This seem to be related to code signing (see the `at MacPackager.signApp` in the error stack trace) which is indeed doomed to fail. 

When I run it from linux code signing is disabled by default (see output below):

```bash
$ release_darwin_dir
  • electron-builder  version=22.5.1 os=4.19.84
  • loaded configuration  file=package.json ("build" field)
  • writing effective config  file=release/builder-effective-config.yaml
  • packaging       platform=darwin arch=x64 electron=7.3.1 appOutDir=release/mac
  • copying Electron  source=/home/rgauthier/dev/gopass-ui/release/electron-dist/Electron.app destination=/home/rgauthier/dev/gopass-ui/release/mac/Electron.app
  • skipped macOS application code signing  reason=supported only on macOS
```

I push another version which sets env var `CSC_IDENTITY_AUTO_DISCOVERY=false` to disable code signing.
According to both <https://github.com/electron-userland/electron-builder/issues/1446> and <https://www.electron.build/code-signing> this should be the way to disable it. Now will it fail to run because not signed, I cannot tell.. Any idea?

If I rely default value for:

> `readonly` = `false` Boolean - Whether to fail if the application is not signed (to prevent unsigned app if code signing configuration is not correct).

excerpt from <https://www.electron.build/configuration/configuration>

it should not fail.

Can you try again? Crossing fingers!

Reply:
Great, build passed now!
What's about running:
* If I run  `result/Applications/Gopass\ UI.app/Contents/MacOS/Gopass\ UI` - nothing happens, but running `result/Applications/Gopass\ UI.app/Contents/MacOS/Gopass\ UI result/Applications/Gopass\ UI.app/Contents/Resources/app.asar` works for me.
* 'bin' and 'share' folders are not needed on darwin:
```
$ ls -la result
result/
├── Applications
│   └── Gopass\ UI.app
...
├── bin
│   └── Gopass\ UI -> ../../pmqygq1wahi9zh83p1fmxmaws75l4zf2-electron-7.3.1/Applications/Electron.app/Contents/MacOS/Electron
└── share
    ├── applications
    │   └── gopass-ui.desktop
    └── icons
...
```

Reply:
Nice to hear. I just pushed another version which uses `wrapProgram`  to ensure the asar is passed as the first arg to electron. Should now be working by just running the exe.

No longer produce `bin` nor `share` on darwin.

I wonder however, will the program be available to be launched from a shell/terminal if not under bin?

Reply:
@sikmir: Any news about this latest version?

Reply:
> I wonder however, will the program be available to be launched from a shell/terminal if not under bin?

No, it will not be available from a shell, but from Launchpad. I see no reason to launch such GUI app from a shell, there is `gopass` for that.

Reply:
Alright, fine by me.

Reply:
Assuming it now work fine on darwin too.